### PR TITLE
[FEATURE] Implemented a PeptideIndexing class

### PR DESCRIPTION
**Pull request by enetz**
Implemented a class with a function equivalent to the tool PeptideIndexer, that updates ProteinIdentification and PeptideIdentification vectors instead of reading and writing IdXML files.
PeptideIndexer also uses this function now and mainly handles the reading and writing of files.

### CODE REVIEW

File: src/openms/include/OpenMS/ANALYSIS/ID/PeptideIndexing.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      UNEXPECTED_RESULT
+    };
+
+    /// Default constructor
+    PeptideIndexing();
+
+    /// Default destructor
+    virtual ~PeptideIndexing();
+
+    /// main method of FeatureFindingMetabo
```
Comment (cbielow) [inline src/openms/include/OpenMS/ANALYSIS/ID/PeptideIndexing.h]: ehmm.. no :)

File: src/topp/PeptideIndexer.cpp [hunk 2]
```diff
@@ -512,14 +173,27 @@ class TOPPPeptideIndexer :
...
 
-    String decoy_string = getStringOption_("decoy_string");
-    bool prefix = getFlag_("prefix");
+
+    PeptideIndexing indexer;
+
+    Param param = getParam_().copy("", true);
+
+    // These parameters are not used by the PeptideIndexing class
+    param.remove("in");
```
Comment (cbielow) [inline src/topp/PeptideIndexer.cpp]: this is not clean and prone to break if new params are added.

Better:
```
Param param_pi = indexer.getParameters();
param_pi.update(param); // overwrite matching params in 'param_pi' with values from 'param'
indexer.setParameters(param);
```

File: src/openms/include/OpenMS/ANALYSIS/ID/PeptideIndexing.h [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+#include <OpenMS/KERNEL/StandardTypes.h>
+#include <OpenMS/CONCEPT/ProgressLogger.h>
+#include <OpenMS/FORMAT/FASTAFile.h>
+
+namespace OpenMS
+{
+
+ class OPENMS_DLLAPI PeptideIndexing :
+    public DefaultParamHandler,
+    public ProgressLogger
```
Comment (cbielow) [inline src/openms/include/OpenMS/ANALYSIS/ID/PeptideIndexing.h]: I don't see it being used (set Progress()...)
nor configured from outside (TOPP tool should forward its config)

File: src/topp/PeptideIndexer.cpp (hunk 2 above)
Comment (timosachsenberg) [inline src/topp/PeptideIndexer.cpp]: agree: update (const Param &old_version, const bool report_new_params=false, const bool only_update_old=false, Logger::LogStream &stream=LOG_WARN)
so I guess update in this case also copies over new ones (as only_update=false is false by default)

File: src/topp/PeptideIndexer.cpp [hunk 4]
```diff
@@ -473,37 +154,17 @@ class TOPPPeptideIndexer :
...
-    registerFlag_("full_tolerant_search", "If set, all peptide sequences are matched using tolerant search. Thus potentially more proteins (containing ambiguous amino acids) are associated. This is much slower!");
-    registerIntOption_("aaa_max", "<number>", 4, "[tolerant search only] Maximal number of ambiguous amino acids (AAA) allowed when matching to a protein database with AAA's. AAA's are 'B', 'Z' and 'X'", false);
-    setMinInt_("aaa_max", 0);
-    registerIntOption_("mismatches_max", "<number>", 0, "[tolerant search only] Maximal number of real mismatches (will be used after checking for ambiguous AA's (see 'aaa_max' option). In general this param should only be changed if you want to look for other potential origins of a peptide which might have unknown SNPs or alike.", false);
-    setMinInt_("mismatches_max", 0);
-    registerFlag_("IL_equivalent", "Treat the isobaric amino acids isoleucine ('I') and leucine ('L') as equivalent (indistinguishable)");
-  }
+//   Param getSubsectionDefaults_(const String& /*section*/) const
```
Comment (timosachsenberg) [inline src/topp/PeptideIndexer.cpp]: remove dead code

File: src/topp/PeptideIndexer.cpp [hunk 5]
```diff
@@ -473,37 +154,17 @@ class TOPPPeptideIndexer :
...
-                                                                                               "\n  '" + EnzymaticDigestion::NamesOfSpecificity[0] + "': both internal cleavage sites must match."
-                                                                                                                                                     "\n  '" + EnzymaticDigestion::NamesOfSpecificity[1] + "': one of two internal cleavage sites must match."
-                                                                                                                                                                                                           "\n  '" + EnzymaticDigestion::NamesOfSpecificity[2] + "': allow all peptide hits no matter their context. Therefore, the enzyme chosen does not play a role here", false);
-    StringList spec;
-    spec.assign(EnzymaticDigestion::NamesOfSpecificity, EnzymaticDigestion::NamesOfSpecificity + EnzymaticDigestion::SIZE_OF_SPECIFICITY);
-    setValidStrings_("enzyme:specificity", spec);
+//    registerSubsection_("algorithm", "Algorithm parameters section");
```
Comment (timosachsenberg) [inline src/topp/PeptideIndexer.cpp]: remove dead code

File: src/topp/PeptideIndexer.cpp [hunk 6]
```diff
@@ -576,457 +248,17 @@ class TOPPPeptideIndexer :
...
-    // annotate target/decoy status of proteins:
-    for (vector<ProteinIdentification>::iterator id_it = prot_ids.begin(); id_it != prot_ids.end(); ++id_it)
+    PeptideIndexing::ExitCodes indexer_exit = indexer.run(proteins, prot_ids, pep_ids);
+    if ( (indexer_exit != PeptideIndexing::EXECUTION_OK) )
     {
-      for (vector<ProteinHit>::iterator hit_it = id_it->getHits().begin(); hit_it != id_it->getHits().end(); ++hit_it)
-      {
-        hit_it->setMetaValue("target_decoy", (protein_is_decoy[hit_it->getAccession()] ? "decoy" : "target"));
-      }
+      return UNKNOWN_ERROR;
```
Comment (timosachsenberg) [inline src/topp/PeptideIndexer.cpp]: maybe also distinguish here UNEXPECTED_RESULT?

**Comment by cbielow** [comment]
Could you also copy the complete help section of the TOPP tool to the class please.
Since it's going to be used internally, the dev's should be able to read up what is does.

I'm not sure how to handle the TOPP tools's params. We could leave them as they are (maintaining a duplicate), but I'm afraid that it will diverge even though its the same code being used.
The alternative is to just put a link to the class in the TOPP tool. But it needs to be prominent, and described as "must read", e.g.

"The tool is described in detail in the PeptideIndexing documentation. Please make sure you read and understand it before using the tool to avoid suboptimal results or long runtimes!"

**Pull request merged**